### PR TITLE
Search & Filtering refactor

### PR DESCRIPTION
**Pull request by SergioMendolia**

### CODE REVIEW

**Comment by SergioMendolia** [comment]
TODO
- [ ] Recreate "unverified" to streamline adding new books
- [ ] remove searchmodal
- [ ] fix tests

**Comment by ragusa87** [comment]
Code-wise, it's a bit hard to follow, and some good practice could be leveraged. I will send some reviews later.

Testing:
- If you change the "sort by" on "all book", it opens the filter panel which is odd.
- By default, there is no results. (All books, series, author), you need to fill some filters to see a book. Looks like a regression to me.
- UX: When you search something like "hacking", the books are displayed behind the filters. But as I don't want to filter anything, I have to dismiss the filter panel every-time to see my search results. I would prefer to have a button "advanced filters" to open it and keep it closed.

File: src/Search/FieldToken.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+        $filtered = [];
+        foreach ($filters as $field => $values) {
+            if ($field === 'orderBy') {
+                continue;
+            }
+            $trimmedfield = trim($field, '-');
+
+            $escapedValues = implode(',', $values);
+            $escapedValues = '['.$escapedValues.']';
```
Comment (ragusa87) [inline src/Search/FieldToken.php:39]: You could do a `json_encode((array)$values)` to avoid the `[` and ` ,` manual handling.

File: src/Search/FieldToken.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+            if (str_ends_with($field, '-')) {
+                $filtered[] = $trimmedfield.':!= '.$escapedValues.' ';
```
Comment (ragusa87) [inline src/Search/FieldToken.php:42]: Avoid if else, use continue.

File: src/Search/QueryTokenizer.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    /**
+     * @param array<TokenInterface> $handlers
+     */
+    public function __construct(
+        #[AutowireIterator('app.search_token')]
+        private iterable $handlers,
+    ) {
+    }
+
+    public function tokenize(string $initialQuery): array
```
Comment (ragusa87) [inline src/Search/QueryTokenizer.php:18]: Please document the return value.

File: src/Search/TokenInterface.php [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace App\Search;
+
+use App\Entity\User;
+use Symfony\Component\DependencyInjection\Attribute\AutoconfigureTag;
+
+#[AutoconfigureTag('app.search_token')]
+interface TokenInterface
+{
```
Comment (ragusa87) [inline src/Search/TokenInterface.php:10]: Se my previous comment to adapt the interface.

File: src/Search/TokenInterface.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    public function getRegex(): string;
+
+    public function setUser(User $user): void;
+
+    public function parseTokens(array $tokens): void;
+
+    public function getFilterQuery(): string;
+
+    public function getOrderQuery(): string;
```
Comment (ragusa87) [inline src/Search/TokenInterface.php:19]: I would add `TEXT` as a constant here and use it.

File: src/Search/TypesenseTokenHandler.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function handle(array $tokens): TypesenseQuery
+    {
+        $query = new TypesenseQuery($tokens['TEXT'] ?? '', 'title,serie,extension,authors,tags,summary');
+
+        $filterstring = [];
+        $orderString = [];
+        foreach ($tokens as $class => $token) {
+            if ($class !== 'TEXT') {
```
Comment (ragusa87) [inline src/Search/TypesenseTokenHandler.php:22]: Avoid long if, do a early-continue instead. Use constant for TEXT.

File: src/Search/TypesenseTokenHandler.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+                $handler = new $class();
```
Comment (ragusa87) [inline src/Search/TypesenseTokenHandler.php:23]: I would instantiate the class after the "instance of" checks, and pass the user in the constructor.

File: src/Service/ShelfManager.php [hunk 8]
```diff
@@ -23,18 +19,19 @@ public function getBooksInShelf(Shelf $shelf): array
...
-        $request = new Request($shelf->getQueryString());
-
-        $this->createAndHandleFilter(BookFilterType::class, $qb, $request);
-
-        $results = $qb->getQuery()->getResult();
-
-        if (!is_array($results)) {
-            return [];
+        $tokens = $this->queryTokenizer->tokenize($shelf->getQueryString());
+        $complexQuery = $this->tokenHandler->handle($tokens);
```
Comment (ragusa87) [inline src/Service/ShelfManager.php:23]: `handle` is too generic,  Use another name, like `$this->tokensToTypesenseQueryAdapter->getQuery($tokens)`

File: src/Twig/Components/InlineEditInteraction.php [hunk 9]
```diff
@@ -71,7 +71,8 @@ public function toggle(): void
         $interaction = $this->getInteraction();
 
         $interaction->setFinished(!$interaction->isFinished());
-
+        $this->book->setUpdated(new \DateTime('now'));
```
Comment (ragusa87) [inline src/Twig/Components/InlineEditInteraction.php:74]: Do you need to set the date ? We do have  `Gedmo\Timestampable` ..

File: src/Twig/Components/InlineEditInteraction.php [hunk 10]
```diff
@@ -91,6 +92,8 @@ public function saveInteraction(): void
         }
 
         $this->entityManager->persist($interaction);
+        $this->book->setUpdated(new \DateTime('now'));
```
Comment (ragusa87) [inline src/Twig/Components/InlineEditInteraction.php:95]: Same regarding `Gedmo\Timestampable`

File: src/Twig/Components/InlineEditInteraction.php [hunk 11]
```diff
@@ -104,6 +107,8 @@ public function toggleFavorite(EntityManagerInterface $entityManager): void
         $interaction->setFavorite(!$interaction->isFavorite());
 
         $entityManager->persist($interaction);
+        $this->book->setUpdated(new \DateTime('now'));
```
Comment (ragusa87) [inline src/Twig/Components/InlineEditInteraction.php:110]: Gedmo\Timestampable

File: src/Twig/Components/InlineEditInteraction.php [hunk 12]
```diff
@@ -116,7 +121,8 @@ public function toggleHidden(EntityManagerInterface $entityManager): void
         $interaction = $this->getInteraction();
 
         $interaction->setHidden(!$interaction->isHidden());
-
+        $this->book->setUpdated(new \DateTime('now'));
```
Comment (ragusa87) [inline src/Twig/Components/InlineEditInteraction.php:124]: Gedmo\Timestampable

File: src/Twig/Components/Search.php [hunk 13]
```diff
@@ -3,64 +3,176 @@
...
 
-    public function __construct(protected CollectionFinder $bookFinder)
+    public function __construct(protected CollectionFinder $bookFinder, protected QueryTokenizer $queryTokenizer, protected TypesenseTokenHandler $tokenHandler, protected Security $security, protected EntityManagerInterface $manager)
     {
     }
 
+    #[PostMount]
+    public function postMount(): void
+    {
+        $tokens = $this->extractFilters(true);
```
Comment (ragusa87) [inline src/Twig/Components/Search.php:57]: It's not possible to understand this first parameter without looking a the method signature.

Either add a method `extractInitialsFilters` or rename to `extractFiltersIsInitial(bool)`

File: src/Twig/Components/Search.php [hunk 14]
```diff
@@ -3,64 +3,176 @@
...
+        $this->getResults($tokens);
+    }
+
     #[LiveAction]
     public function addToQuery(#[LiveArg] string $value): void
```
Comment (ragusa87) [inline src/Twig/Components/Search.php:62]: This method do way more than `addToQuery` (alter the filters, extract token, etc), "adding" feels confusing,  rename to something else like `alterQuery`

File: src/Twig/Components/Search.php [hunk 15]
```diff
@@ -3,64 +3,176 @@
...
+        $results = $result->getResults();
+        $this->facets = $result->getFacetCounts();
 
-        foreach ($rawResults as $result) {
-            $document = $result['document'];
-            foreach ($result['highlights'] as $highlight) {
-                $document[$highlight['field']] = $highlight['snippet'] ?? '';
-            }
-            $this->books[$document['id']] = $document;
+        foreach (['read', 'hidden', 'favorite'] as $item) {
```
Comment (ragusa87) [inline src/Twig/Components/Search.php:121]: As mentioned before, use constants.

File: src/Twig/Components/Search.php [hunk 16]
```diff
@@ -3,64 +3,176 @@
...
+        if (($tokens['TEXT'] ?? '') !== $this->query) {
+            $this->query = $tokens['TEXT'] ?? '';
         }
 
-        $complexQuery = new TypesenseQuery($this->query, 'title,authors,serie,tags,summary,serieIndex');
+        return $tokens;
+    }
 
-        $complexQuery->facetBy('authors,serie,tags');
+    protected function getResults(array $tokens): void
```
Comment (ragusa87) [inline src/Twig/Components/Search.php:107]: Rename to `processResults` it's not a getter (return void).

File: .github/workflows/tests.yml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+      - name: Install NPM Dependencies
+        run: docker compose exec biblioteca npm i
+
+      - name: Build npm
+        run: docker compose exec biblioteca npm run build
+
+      - name: Prepare Database
+        run: |
+          docker compose exec biblioteca bin/console doctrine:database:create --env=test -n
+          docker compose exec biblioteca bin/console doctrine:schema:update --complete --force --env=test -n
```
Comment (ragusa87) [inline .github/workflows/tests.yml:46]: Don't use `doctrine:schema:update`, but the migration instead.

> php bin/console doctrine:migrations:migrate --env=test -n

File: .env.test [hunk 18]
```diff
@@ -5,12 +5,11 @@ SYMFONY_DEPRECATIONS_HELPER=999999
...
-
-TYPESENSE_URL=http://localhost:8108
+TYPESENSE_URL=http://typesense:8108
 TYPESENSE_KEY=xyz
 
 # Disable the proxy for the test env
 KOBO_PROXY_USE_EVERYWHERE=0
 KOBO_PROXY_ENABLED=0
+
+DATABASE_URL="mysql://root:biblioteca@db:3306/biblioteca?serverVersion=mariadb-10.3.39&charset=utf8"
```
Comment (ragusa87) [inline .env.test:15]: The `serverVersion` format changed with DBAL 4.x, use `serverVersion=10.3.39-MariaDB`
See https://symfony.com/doc/current/reference/configuration/doctrine.html

File: src/Twig/Components/InlineEditInteraction.php (hunk 9 above)
Comment (SergioMendolia) [reply src/Twig/Components/InlineEditInteraction.php:74]: That's because of TypesenseBundle... I want to reindex the document when I change an interaction, but interaction is another entity so the book is not updated, therefore not indexed. If we had a bundle that allowed to reindex the book without changing it it wouldn't be necessary.

File: migrations/Version20241223182304.php [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    public function getDescription(): string
+    {
+        return 'Store typesense query instead of json';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE shelf CHANGE query_string query_string LONGTEXT DEFAULT NULL');
+    }
```
[the remark this answers is not in the document]
Comment (SergioMendolia) [reply migrations/Version20241223182304.php:24]: no... it will not work properly as it will just put json in the search field but I think it's still better than just losing all. the users can then see how it was ocnfigured and adapt it

File: src/Entity/Book.php [hunk 20]
```diff
@@ -587,4 +587,32 @@ public function setBookmarkUsers(Collection $bookmarkUsers): self
...
+        ];
+
+        foreach ($this->getBookInteractions() as $interaction) {
+            $user = $interaction->getUser();
+            if ($user === null) {
+                continue;
+            }
+            $userId = $user->getId();
+            if ($interaction->isFinished()) {
+                $return['read'][] = $userId;
```
[the remark this answers is not in the document]
Comment (SergioMendolia) [reply src/Entity/Book.php:606]: Treating this in another MR, this part is very temporary
